### PR TITLE
sg: add an auto update mechanism

### LINKED ISSUE #29619: sg: prompt the user for auto-updating rather than issuing a warning that new version is available.
Quoting @unknwon from Slack: 

> I'm doing it almost every single day... and I need to open a separate tab for running this command... could we instead prompt a "Y/n" (default no) and do the upgrade automatically (like oh-my-zsh) :grin:

### PR DESCRIPTION
**Pull request by jhchabran**
Introduces a new subcommand `sg update auto enable` to turn on automatic updates. It's turned off by default. 

When turned on, `sg` will first check in its local copy of `origin/main` if there are new commits. If yes, it will execute `sg update` and restart itself with the same arguments, transparently running the given arguments to the freshly updated `sg` and requiring no action from the user. 

CleanShot 2022-03-15 at 12 54 23

Fixes #29619

## Test plan

1. Replaced the update check with `main` instead of `origin/main`
1. Manually compiled `sg` with my changes 
1. Ran `sg update auto enable`
1. Ran `sg ci preview` 
    - Got an update triggered
    - Still got the output with from `sg ci preview`

See the gif in the above description.

### CODE REVIEW

File: dev/sg/main.go [hunk 1]
```diff
@@ -144,7 +160,12 @@ func main() {
 		os.Exit(1)
 	}
 
-	checkSgVersion()
+	err := checkSgVersion(ctx)
+	if err != nil {
+		fmt.Printf("error: %s\n", err)
```
Comment (mrnugget) [inline dev/sg/main.go:165]:
```suggestion
		fmt.Printf("checking sg version failed: %s\n", err)
```

File: dev/sg/main.go [hunk 2]
```diff
@@ -117,11 +117,27 @@ func checkSgVersion() {
...
-		stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "             To see what's new, run 'sg version changelog -next'.             "))
-		stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "------------------------------------------------------------------------------"))
+		if !getAutoUpdateSetting(ctx) {
+			stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "------------------------------------------------------------------------------"))
+			stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "       HEY! New version of sg available. Run 'sg update' to install it.       "))
+			stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "             To see what's new, run 'sg version changelog -next'.             "))
+			stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "------------------------------------------------------------------------------"))
+		} else {
+			stdout.Out.WriteLine(output.Line(output.EmojiInfo, output.StyleSuggestion, "Auto updating sg ..."))
+			err := updateCommand.Exec(context.TODO(), nil)
```
Comment (mrnugget) [inline dev/sg/main.go:127]: Or am I missing something?

```suggestion
			err := updateCommand.Exec(ctx, nil)
```

File: dev/sg/main.go [hunk 3]
```diff
@@ -117,11 +117,27 @@ func checkSgVersion() {
 
 	out = strings.TrimSpace(out)
 	if out != "" {
```
Comment (mrnugget) [inline dev/sg/main.go:119]: Maybe now's the time to turn this into an early-return:

```go
if out == "" {
  // No newer commits found. sg is up to date.
  return nil
}

File: dev/sg/main.go [hunk 4]
```diff
@@ -117,11 +117,27 @@ func checkSgVersion() {
...
+			err := updateCommand.Exec(context.TODO(), nil)
+			if err != nil {
+				return err
+			}
+			sgPath, err := os.Executable()
+			if err != nil {
+				return err
+			}
+			if err := syscall.Exec(sgPath, os.Args, os.Environ()); err != nil {
+				return err
```
Comment (mrnugget) [inline dev/sg/main.go:136]:
```suggestion
			return syscall.Exec(sgPath, os.Args, os.Environ())
```

**Review by mrnugget — COMMENTED** [review]
Nice! Thanks for picking this up!

This is subjective, but: I do think it would be easier for now if we manage `auto-update: true/false` via a command line flag. Not only because it doesn't require checking state, but mainly because I think it's far too easy to disable it for a specific reason (you're working on a branch, want to debug something, etc. and don't want to update) and then forgetting about it.

So if I'd implement this I'd make auto-updates the default (yes, let's do it!) and add a global `-no-auto-update` flag that, for that invocation of `sg`, would skip auto-updating.

File: dev/sg/main.go (hunk 2 above)
Comment (jhchabran) [reply dev/sg/main.go:127]: Woops, thanks.

File: dev/sg/main.go [hunk 5]
```diff
@@ -144,7 +160,12 @@ func main() {
 		os.Exit(1)
 	}
 
-	checkSgVersion()
+	err := checkSgVersion(ctx)
```
Comment (mrnugget) [inline dev/sg/main.go:163]: While we're here: maybe we can do a check here that skips this if we're running `sg update`:

```go
if !(len(os.Args) >= 2 && os.Args[1] == "update") {
  err := checkSgVersion(ctx)
}
```

Or whatever the right combination of boolean conditions is. I realised yesterday I struggle when it's more than two booleans.

**Comment by jhchabran** [comment]
> Nice! Thanks for picking this up!
> 
> This is subjective, but: I do think it would be easier for now if we manage `auto-update: true/false` via a command line flag. Not only because it doesn't require checking state, but mainly because I think it's far too easy to disable it for a specific reason (you're working on a branch, want to debug something, etc. and don't want to update) and then forgetting about it.

@mrnugget Just so I understand your point correctly, are you talking about working *on `sg`* in a branch or working on something else? 

If that's the former, `sg` won't update, because your version is going to be prefixed by `dev-` and the update mechanism will see that and just bail (silently though, we may want to make it explicit).  

If it's the latter, I'm not sure to see how it would disrupt your flow, minus the few seconds to build it.

File: dev/sg/main.go (hunk 5 above)
Comment (jhchabran) [reply dev/sg/main.go:163]: Oh right, good catch

**Comment by mrnugget** [comment]
> @mrnugget Just so I understand your point correctly, are you talking about working _on `sg`_ in a branch or working on something else?
> 
> If that's the former, `sg` won't update, because your version is going to be prefixed by `dev-` and the update mechanism will see that and just bail (silently though, we may want to make it explicit).
> 
> If it's the latter, I'm not sure to see how it would disrupt your flow, minus the few seconds to build it.

Ah, you're right! I guess I was thinking of a made-up usecase here where I'm working/git-bisecting something in a branch and want to run it with the `sg` version for that commit. But that's independent of this feature.

That being said, I do think there's a bigger case for making "disable auto-updates" a non-permanent setting. Here it is: we do want to make auto-updates the default (again: I think we can/should do it now). Once it's enabled-by-default I don't see a usecase for disabling it permanently.

File: dev/sg/main.go (hunk 5 above)
Comment (jhchabran) [reply dev/sg/main.go:163]: Damn, my brain can't process your boolean expression without thinking it's wrong 🤣

**Comment by mrnugget** [comment]
> My approach here is to ship an opt-in setting, then to see if we aren't seeing any issues for a few days, then we can make it the default setting. WDYT?

Yeah, that makes sense, but in this case it's so easy to revert and we know the complete userbase by their first name, and everybody's been typing `./dev/sg/install.sh && sg` for a while (some of us have aliases!) that I think the danger is low :)

**Comment by jhchabran** [comment]
> > My approach here is to ship an opt-in setting, then to see if we aren't seeing any issues for a few days, then we can make it the default setting. WDYT?
> 
> Yeah, that makes sense, but in this case it's so easy to revert and we know the complete userbase by their first name, and everybody's been typing `./dev/sg/install.sh && sg` for a while (some of us have aliases!) that I think the danger is low :)

Ok, let's go for it :D

File: dev/sg/main.go [hunk 6]
```diff
@@ -116,12 +116,29 @@ func checkSgVersion() {
...
+		// No newer commits found. sg is up to date.
+		return nil
+	}
+
+	if *skipAutoUpdatesFlag {
 		stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "------------------------------------------------------------------------------"))
 		stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "       HEY! New version of sg available. Run 'sg update' to install it.       "))
 		stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "             To see what's new, run 'sg version changelog -next'.             "))
 		stdout.Out.WriteLine(output.Linef("", output.StyleSearchMatch, "------------------------------------------------------------------------------"))
+	} else {
```
Comment (mrnugget) [inline dev/sg/main.go:129]: Since the other branch returns, this might be clearer:

```suggestion
		return nil
	}
```

**Review by mrnugget — APPROVED** [review]

**Pull request merged**